transformation/src/main/scala/org/broadinstitute/monster/encode/transformation/TransformationPipelineBuilder.scala

```diff
@@ -180,6 +179,82 @@ object TransformationPipelineBuilder extends PipelineBuilder[Args] {
         .groupByKey
     }
 
+    // Get analysis step objects
+    val analysisStepRuns = readRawEntities(EncodeEntity.AnalysisStepRun)
+    val analysisStepVersionsById = readRawEntities(EncodeEntity.AnalysisStepVersion)
+      .withName("Key analysis step versions by ID")
+      .keyBy(_.read[String]("@id"))
+    val analysisStepsById = readRawEntities(EncodeEntity.AnalysisStep)
+      .withName("Key analysis steps by ID")
+      .keyBy(_.read[String]("@id"))
+    val filesByStepRun = fileInputs
+      .withName("Key files by step run ID")
+      .keyBy(_.tryRead[String]("step_run").getOrElse(""))
+      .groupByKey
+
+    // Join AnalysisStepRuns, AnalysisStepVersions, AnalysisSteps, and Files.
+    // They will be used in both the StepRun and PipelineRun transformations.
+    val stepRunInfo = analysisStepRuns
+      .withName("Key step runs by analysis step version")
+      .keyBy(_.read[String]("analysis_step_version"))
+      .join(analysisStepVersionsById)
+      .values
+      .withName("Key step runs by analysis step")
+      .keyBy(_._2.read[String]("analysis_step"))
+      .join(analysisStepsById)
+      .values // outputs ((stepRun, stepVersion), step)
+      .withName("Key step runs by step run ID")
+      .keyBy(_._1._1.read[String]("@id"))
+      .leftOuterJoin(filesByStepRun)
+      .values
+      .withSideInputs(fileIdToType)
+      .withName("Transform step runs")
+
+    // Transform step runs
+    val stepRunOutput = stepRunInfo.map {
+      case ((((stepRun, stepVersion), step), generatedFiles), sideCtx) =>
```

Comment:
I tried making a `StepRunInfo` case class to use instead of this messy tuple but dataflow complained that it wasn't serializable. It looks like there are ways to make it serializable but I don't know that it's worth it: 
https://stackoverflow.com/questions/28032063/how-to-fix-dataflow-unable-to-serialize-my-dofn
Let me know if anyone has thoughts or opinions on it.

Reply:
Huh, it should Just Work as long as the individual fields are serializable.

I don't feel strongly about having a `case class` now that the tuple-based form is working. I think if we ever need to touch this code again, we might want to reconsider it.